stakingModel/contracts/GoodReserveCDai.sol

```diff
@@ -62,40 +67,145 @@ contract GoodReserveCDai is DSMath, SchemeGuard {
+            "You need to approve cDAI transfer first"
+        );
+        require(
+            buyWith.transferFrom(msg.sender, address(this), tokenAmount) == true,
+            "transferFrom failed, make sure you approved cDAI transfer"
+        );
+        uint256 gdReturn = marketMaker.buy(buyWith, tokenAmount);
+        require(gdReturn >= minReturn, "GD return must be above the minReturn");
+        ERC20Mintable(address(gooddollar)).mint(msg.sender, gdReturn);
+        emit TokenPurchased(msg.sender, address(buyWith), tokenAmount, minReturn, gdReturn);
+        return gdReturn;
+    }
+
+    /**
+    * @dev calculate the deducted amount during the sell action. there is a
+    * `sellDeductionRatio` percent deduction
+    * @return (deductedAmount) the deducted amount for sell
     */
-    function buy(ERC20 buyWith, uint256 tokenAmount, uint256 maxPrice)
+    function calculateSellDeduction(uint256 gdAmount)
         public
-        payable
-    {}
+        view
+        returns (uint256)
+    {
+
+        uint256 decimalsDiff = uint256(27).sub(uint256(gooddollar.decimals()));
+        uint256 deduction =
+        rmul(
+                gdAmount.mul(10**decimalsDiff), // expand to e27 precision
+                sellDeductionRatio
+            )
+                .div(10**decimalsDiff); // return to e2 precision
+        require(gdAmount > deduction, "Calculation error");
+        return gdAmount.sub(deduction);
+    }
 
     /**
-    * @dev Sell G$ to the marketMaker
-    * currently only cDai is supported
+    * @dev sell G$ to sellTo and update the bonding curve params. sell occurs only if the
+    * token return is above the given minimum. notice that there is a deducted
+    * amount from the given G$ that stays in the reserve. it is possible to sell only to
+    * cDAI and when the contract is set to active. MUST call to G$ `approve` prior this
+    * selling action to allow this contract to accomplish the conversion.
+    * @param gdAmount how much G$ tokens convert to `sellTo` tokens
+    * @param minReturn the minimum allowed return in `sellTo` tokens
+    * @return (tokenReturn) how much `sellTo` tokens were transferred
     */
-    function sell(ERC20 sellTo, uint256 gdAmount, uint256 minPrice) public {}
+    function sell(ERC20 sellTo, uint256 gdAmount, uint256 minReturn)
+        public
+        active
+        onlyCDai(sellTo)
+        returns (uint256)
+    {
+        ERC20Burnable(address(gooddollar)).burnFrom(msg.sender, gdAmount);
+        uint256 deductedGdAmount = calculateSellDeduction(gdAmount);
+        uint256 tokenReturn = marketMaker.sellWithDeduction(sellTo, gdAmount, deductedGdAmount);
+        require(tokenReturn >= minReturn, "Token return must be above the minReturn");
+        require(sellTo.transfer(msg.sender, tokenReturn) == true, "Transfer failed");
+        emit TokenSold(msg.sender, address(sellTo), gdAmount, minReturn, tokenReturn);
```

Comment:
maybe add the deducted amount to emitted event? call it "contribution" and not deduction
do you think minReturn should be calculated before or after deduction?

Reply:
I think after